netty/src/main/java/io/grpc/netty/WriteCombiningHandler.java

```diff
@@ -0,0 +1,153 @@
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package io.grpc.netty;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.netty.util.ReferenceCountUtil.safeRelease;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.EmptyByteBuf;
+import io.netty.buffer.UnpooledByteBufAllocator;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelOutboundHandlerAdapter;
+import io.netty.channel.ChannelPromise;
+import io.netty.channel.DefaultChannelPromise;
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.FutureListener;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class WriteCombiningHandler extends ChannelOutboundHandlerAdapter {
+
+  private static final int INITIAL_BUFFER_SIZE = 4096;
+  private static final int MAX_COPY_BUFFER_SIZE = 256;
+
+  private ByteBuf buffer;
+  private CollectivePromise bufferPromise;
+
+  private ChannelHandlerContext ctx;
+
+  @Override
+  public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+    buffer = ctx.alloc().directBuffer(INITIAL_BUFFER_SIZE, INITIAL_BUFFER_SIZE);
+    bufferPromise = new CollectivePromise(ctx);
+    this.ctx = ctx;
+  }
+
+  @Override
+  public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+    safeRelease(buffer);
+    buffer = null;
+    bufferPromise = null;
+    this.ctx = null;
+  }
+
+  @Override
+  public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+    if (!(msg instanceof ByteBuf)) {
+      writeBuffer();
+      ctx.write(msg, promise);
+      return;
+    }
+    final ByteBuf data = (ByteBuf) msg;
+    if (data.readableBytes() > MAX_COPY_BUFFER_SIZE) {
+      writeBuffer();
+      ctx.write(data, promise);
+    } else if (data.readableBytes() > buffer.writableBytes()) {
+      writeBuffer();
+      copyBytes(data, promise);
+    } else {
+      copyBytes(data, promise);
+    }
+  }
```

Comment:
@buchgr I think you also want to override `close(...)` and `disconnect(...)` and ensure that the buffer is written before you process these.